clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1716,9 +1737,18 @@ class SyclKernelUnionChecker : public SyclKernelFieldHandler {
   }
 };
 
-// A type to mark whether a collection requires decomposition.
+// A type to mark whether a collection requires decomposition
+// or needs to be transformed to a new type. If a collection
+// contains pointers, and is not decomposed, a new type must
+// be generated with all pointers in global address space.
 class SyclKernelDecompMarker : public SyclKernelFieldHandler {
   llvm::SmallVector<bool, 16> CollectionStack;
+  llvm::SmallVector<bool, 16> PointerStack;
+
+  // TODO: Remove this method once arrays are handled correctly.
```

Comment:
Think this can be removed now since I changed logic. Will do so in next push with any other review comments.